src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Recognition/Precompute.hs

```diff
@@ -79,85 +81,74 @@ import Swarm.Util (histogram)
 
   process g = StructureInfo g entGrid countsMap
    where
-    entGrid = getEntityGrid extractor $ namedGrid g
-    countsMap = histogram $ concatMap catMaybes entGrid
-
-  infos =
-    M.fromList $
-      map (name . namedGrid &&& process) xs
+    entGrid = extractedGrid $ grid g
+    countsMap = histogram . catMaybes . NE.toList $ allMembersNE entGrid
 
--- | Returns 'Nothing' if the grid is empty, since we want
--- to exclude them from the recognition engine.
 extractOrientedGrid ::
-  (Maybe b -> Maybe a) ->
-  NamedGrid (Maybe b) ->
+  ExtractedArea b a ->
   AbsoluteDir ->
-  Maybe (StructureWithGrid (Maybe b) a)
-extractOrientedGrid extractor x d =
-  case extractor <$> structure x of
-    EmptyGrid -> Nothing
-    Grid neGrid ->
-      let w = RowWidth . rectWidth . getNEGridDimensions $ neGrid
-       in Just $
-            StructureWithGrid wrapped d w $
-              applyOrientationTransformNE (Orientation d False) neGrid
+  StructureWithGrid b a
+extractOrientedGrid (ExtractedArea x neGrid) d =
+  StructureWithGrid d w $
+    ExtractedArea x $
+      applyOrientationTransformNE (Orientation d False) neGrid
  where
-  wrapped = NamedOriginal (name x) x
+  w = RowWidth . rectWidth . getNEGridDimensions $ neGrid
 
 -- |
 -- At this point, we have already ensured that orientations
 -- redundant by rotational symmetry have been excluded
 -- (i.e. at Scenario validation time).
---
--- Excludes empty grids.
 extractGrids ::
-  (Maybe b -> Maybe a) ->
-  NamedGrid (Maybe b) ->
-  [StructureWithGrid (Maybe b) a]
-extractGrids extractor x =
-  mapMaybe (extractOrientedGrid extractor x) orientations
+  ExtractedArea b a ->
+  [StructureWithGrid b a]
+extractGrids x =
+  map (extractOrientedGrid x) orientations
  where
-  orientations = Set.toList $ recognize x
+  orientations = Set.toList $ recognize $ originalItem x
 
 -- | The output list of 'FoundStructure' records is not yet
 -- vetted; the 'ensureStructureIntact' function will subsequently
 -- filter this list.
 lookupStaticPlacements ::
-  (Maybe b -> Maybe a) ->
-  StaticStructureInfo b ->
-  [FoundStructure (Maybe b) a]
-lookupStaticPlacements extractor (StaticStructureInfo structDefs thePlacements) =
+  StaticStructureInfo b a ->
+  [FoundStructure b a]
+lookupStaticPlacements (StaticStructureInfo theAutomatons thePlacements) =
   concatMap f $ M.toList thePlacements
  where
-  definitionMap = M.fromList $ map ((name &&& id) . namedGrid) structDefs
+  definitionMap = theAutomatons ^. originalStructureDefinitions
 
   f (subworldName, locatedList) = mapMaybe g locatedList
    where
     g (LocatedStructure (OrientedStructure theName d) loc) = do
       sGrid <- M.lookup theName definitionMap
-      x <- extractOrientedGrid extractor sGrid d
-      return $ PositionedStructure (Cosmic subworldName loc) x
+      return $
+        PositionedStructure (Cosmic subworldName loc) $
+          extractOrientedGrid (grid $ annotatedGrid sGrid) d
 
```

Comment:
This was the other place that "extraction" was recomputed.  Now we just re-use the result computed earlier.